texar/evals/bleu_tool.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2019 The Texar Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Python implementation of BLEU adapted from:
+    `https://github.com/tensorflow/models/blob/master/official/transformer/compute_bleu.py`
+"""
+
+from typing import Counter, List, Tuple
+
+import re
+import sys
+import unicodedata
+import collections
+import math
+import numpy as np
+
+__all__ = [
+    "compute_bleu",
+    "bleu_tokenize",
+    "bleu_wrapper",
+]
+
+
+def _get_ngrams(segment: List[str],
+                max_order: int) -> Counter[Tuple[str, ...]]:
+    r"""Extracts all n-grams up to a given maximum order from an
+    input segment.
+
+    Args:
+        segment: text segment from which n-grams will be extracted.
+        max_order: maximum length in tokens of the n-grams returned
+            by this methods.
+
+    Returns:
+        The Counter containing all n-grams upto :attr:`max_order`
+        in segment with a count of how many times each n-gram occurred.
+    """
+    ngram_counts: collections.Counter = collections.Counter()
+    for order in range(1, max_order + 1):
+        for i in range(0, len(segment) - order + 1):
+            ngram = tuple(segment[i:i + order])
+            ngram_counts[ngram] += 1
+    return ngram_counts
+
+
+def compute_bleu(reference_corpus: List[List[str]],
+                 translation_corpus: List[List[str]],
+                 max_order: int = 4,
+                 use_bp: bool = True) -> float:
+    r"""Computes BLEU score of translated segments against references.
```

Comment:
the name `compute_bleu` is a bit too vague. Let's call it `corpus_bleu_transformer` (on par w/ `corpus_bleu_moses` that indicates the version of BLEU "moses")